vumi/persist/tests/test_fake_redis.py

```diff
@@ -1,491 +1,750 @@
+        self.assert_redis_op(redis, ['31', result_keys[5:15]], 'scan', '10')
+        self.assert_redis_op(redis, [None, result_keys[15:]], 'scan', '31')
 
-    @inlineCallbacks
-    def test_scan_no_keys(self):
-        self.assertEqual(
-            (yield self.redis.scan(None)),
-            (None, []))
 
+class TestFakeRedis(FakeRedisUnverifiedTestMixin, FakeRedisTestMixin,
+                    VumiTestCase):
+
+    def get_redis(self, **kwargs):
+        redis = FakeRedis(**kwargs)
+        self.add_cleanup(redis.teardown)
+        return redis
+
+    def assert_redis_op(self, redis, expected, op, *args, **kw):
+        self.assertEqual(expected, getattr(redis, op)(*args, **kw))
+
+    def assert_redis_error(self, redis, op, *args, **kw):
+        self.assertRaises(Exception, getattr(redis, op), *args, **kw)
 
-class TestFakeRedisCharsetHandling(VumiTestCase):
+    def wait(self, redis, delay):
+        redis.clock.advance(delay)
 
-    def get_redis(self, *args, **kwargs):
-        redis = FakeRedis(*args, **kwargs)
+
+class TestFakeRedisAsync(FakeRedisUnverifiedTestMixin, FakeRedisTestMixin,
+                         VumiTestCase):
+
+    def get_redis(self, **kwargs):
+        redis = FakeRedis(async=True, **kwargs)
         self.add_cleanup(redis.teardown)
         return redis
 
     def assert_redis_op(self, redis, expected, op, *args, **kw):
+        d = getattr(redis, op)(*args, **kw)
+        return d.addCallback(lambda r: self.assertEqual(expected, r))
+
+    def assert_redis_error(self, redis, op, *args, **kw):
+        d = getattr(redis, op)(*args, **kw)
+        return self.assertFailure(d, Exception)
+
+    def wait(self, redis, delay):
+        redis.clock.advance(delay)
+
+
+class RedisPairWrapper(object):
+    def __init__(self, test_case, fake_redis, real_redis):
+        self._test_case = test_case
+        self._fake_redis = fake_redis
+        self._real_redis = real_redis
+        self._perform_operation = self._real_redis.call_decorator(
+            self._perform_operation_gen)
+
+    def _perform_operation_gen(self, op, *args, **kw):
+        """
+        Perform an operation on both the fake and real Redises and assert that
+        the responses and errors are the same.
+
+        NOTE: This method is a generator and is not used directly. It's wrapped
+              with an appropriate sync/async wrapper in __init__() above.
+        """
+        results = []
+        errors = []
+        for redis in [self._fake_redis, self._real_redis]:
+            try:
+                result = yield getattr(redis, op)(*args, **kw)
+            except Exception as e:
+                errors.append(e)
+            else:
+                results.append(result)
+
+        # First, handle errors.
+        if errors:
+            if results:
+                self._test_case.fail(
+                    "One redis returned %r and the other raised %r" % (
+                        results[0], errors[0]))
```

Comment:
It would be helpful to know which one returned something and which one raised something, might save us some confusion.